src/Symfony/Component/HttpClient/RecordReplayCallback.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+/*
+ *  This file is part of the Symfony package.
+ *
+ *  (c) Fabien Potencier <fabien@symfony.com>
+ *
+ *  For the full copyright and license information, please view the LICENSE
+ *  file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerAwareTrait;
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Component\HttpClient\Response\MockResponse;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * A callback for the MockHttpClient. Three modes available:
+ * - MODE_RECORD -> Make an actual HTTP request and save the response, overriding any pre-existing response.
+ * - MODE_REPLAY -> Will try to replay an existing response, and throw an exception if none is found
+ * - MODE_REPLAY_OR_RECORD -> Try to replay response if possible, otherwise make an actual HTTP request and save it.
+ *
+ * @author Gary PEGEOT <gary.pegeot@allopneus.com>
+ */
+class RecordReplayCallback implements LoggerAwareInterface
+{
+    use HttpClientTrait;
+    use LoggerAwareTrait;
+
+    public const MODE_RECORD = 'record';
+    public const MODE_REPLAY = 'replay';
+    public const MODE_REPLAY_OR_RECORD = 'replay_or_record';
+
+    /**
+     * @var string
+     */
+    private $mode;
+
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cache;
+
+    /**
+     * @var HttpClientInterface
+     */
+    private $client;
+
+    public function __construct(string $mode, CacheItemPoolInterface $cache, HttpClientInterface $client = null)
```

Comment:
Shouldn't we make the mode optional, defaulting to MODE_REPLAY_OR_RECORD?
That would ease integration with tests I believe: one wants to re-record? => `bin/console cache:pool:clear some_pool_for_some_reqs` then run the tests again.
WDYT?